File: Y2/POO/lab11/decorator_ex.cpp
#include <iostream>
#include <string>
using namespace std;

class Shaorma {
public:
    virtual string getDescription() = 0;
    virtual double getCost() = 0;
};

class EmptyShaorma : public Shaorma {
public:
    string getDescription() override {
        return "Empty Shaorma";
    }

    double getCost() override {
        return 0;
    }
};

class ShaormaDecorator : public Shaorma {
protected:
    Shaorma* decoratedShaorma;

public:
    ShaormaDecorator(Shaorma* decoratedShaorma) : decoratedShaorma(decoratedShaorma) {}

    string getDescription() override {
        return decoratedShaorma->getDescription();
    }

    double getCost() override {
        return decoratedShaorma->getCost();
    }
};

class PuiDecorator : public ShaormaDecorator {
public:
    PuiDecorator(Shaorma* decoratedShaorma) : ShaormaDecorator(decoratedShaorma) {}

    string getDescription() override {
        return decoratedShaorma->getDescription() + ", Pui";
    }

    double getCost() override {
        return decoratedShaorma->getCost() + 5;
    }
};

class VitaDecorator : public ShaormaDecorator {
public:
    VitaDecorator(Shaorma* decoratedShaorma) : ShaormaDecorator(decoratedShaorma) {}

    string getDescription() override {
        return decoratedShaorma->getDescription() + ", Vita";
    }

    double getCost() override {
        return decoratedShaorma->getCost() + 9;
    }
};

class VarzaDecorator : public ShaormaDecorator {
public:
    VarzaDecorator(Shaorma* decoratedShaorma) : ShaormaDecorator(decoratedShaorma) {}

    string getDescription() override {
        return decoratedShaorma->getDescription() + ", Varza";
    }

    double getCost() override {
        return decoratedShaorma->getCost() + 3;
    }
};

class KetchupDecorator : public ShaormaDecorator {
public:
    KetchupDecorator(Shaorma* decoratedShaorma) : ShaormaDecorator(decoratedShaorma) {}

    string getDescription() override {
        return decoratedShaorma->getDescription() + ", Ketchup";
    }

    double getCost() override {
        return decoratedShaorma->getCost() + 9;
    }
};

class CastravetiDecorator : public ShaormaDecorator {
public:
    CastravetiDecorator(Shaorma* decoratedShaorma) : ShaormaDecorator(decoratedShaorma) {}

    string getDescription() override {
        return decoratedShaorma->getDescription() + ", Castraveti";
    }

    double getCost() override {
        return decoratedShaorma->getCost() + 9;
    }
};
int main() {
    // Empty Shaorma
    Shaorma* empty = new EmptyShaorma();
    cout << "Description: " << empty->getDescription() << endl;
    cout << "Cost: $" << empty->getCost() << endl;

    // Shaorma with Milk
    Shaorma* shaormaCuVita = new VitaDecorator(new VarzaDecorator(new EmptyShaorma()));
    cout << "\nDescription: " << shaormaCuVita->getDescription() << endl;
    cout << "Cost: $" << shaormaCuVita->getCost() << endl;

    // Shaorma cu de toate
    Shaorma* shaormaCuDeToate = new VitaDecorator(new VarzaDecorator(new PuiDecorator(new CastravetiDecorator(new KetchupDecorator(new EmptyShaorma())))));
    cout << "\nDescription: " << shaormaCuDeToate->getDescription() << endl;
    cout << "Cost: $" << shaormaCuDeToate->getCost() << endl;

    delete empty;
    delete shaormaCuVita;
    delete shaormaCuDeToate;
    return 0;
}

File: Y2/POO/lab11/kotlin/decorator_ex.kt
abstract class Shaorma {
    abstract fun getDescription(): String
    abstract fun getCost(): Double
}

class EmptyShaorma : Shaorma() {
    override fun getDescription(): String {
        return "Empty Shaorma"
    }

    override fun getCost(): Double {
        return 0.0
    }
}

open class ShaormaDecorator(protected val decoratedShaorma: Shaorma) : Shaorma() {
    override fun getDescription(): String {
        return decoratedShaorma.getDescription()
    }

    override fun getCost(): Double {
        return decoratedShaorma.getCost()
    }
}

class PuiDecorator(decoratedShaorma: Shaorma) : ShaormaDecorator(decoratedShaorma) {
    override fun getDescription(): String {
        return decoratedShaorma.getDescription() + ", Pui"
    }

    override fun getCost(): Double {
        return decoratedShaorma.getCost() + 5
    }
}

class VitaDecorator(decoratedShaorma: Shaorma) : ShaormaDecorator(decoratedShaorma) {
    override fun getDescription(): String {
        return decoratedShaorma.getDescription() + ", Vita"
    }

    override fun getCost(): Double {
        return decoratedShaorma.getCost() + 9
    }
}

class VarzaDecorator(decoratedShaorma: Shaorma) : ShaormaDecorator(decoratedShaorma) {
    override fun getDescription(): String {
        return decoratedShaorma.getDescription() + ", Varza"
    }

    override fun getCost(): Double {
        return decoratedShaorma.getCost() + 3
    }
}

class KetchupDecorator(decoratedShaorma: Shaorma) : ShaormaDecorator(decoratedShaorma) {
    override fun getDescription(): String {
        return decoratedShaorma.getDescription() + ", Ketchup"
    }

    override fun getCost(): Double {
        return decoratedShaorma.getCost() + 9
    }
}

class CastravetiDecorator(decoratedShaorma: Shaorma) : ShaormaDecorator(decoratedShaorma) {
    override fun getDescription(): String {
        return decoratedShaorma.getDescription() + ", Castraveti"
    }

    override fun getCost(): Double {
        return decoratedShaorma.getCost() + 9
    }
}

fun main() {
    val empty = EmptyShaorma()
    println("Description: ${empty.getDescription()}")
    println("Cost: $${empty.getCost()}")

    val shaormaCuVita = VitaDecorator(VarzaDecorator(EmptyShaorma()))
    println("\nDescription: ${shaormaCuVita.getDescription()}")
    println("Cost: $${shaormaCuVita.getCost()}")

    val shaormaCuDeToate = VitaDecorator(
        VarzaDecorator(
            PuiDecorator(
                CastravetiDecorator(
                    KetchupDecorator(EmptyShaorma())
                )
            )
        )
    )
    println("\nDescription: ${shaormaCuDeToate.getDescription()}")
    println("Cost: $${shaormaCuDeToate.getCost()}")
}
